service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/storageContainer/CreateAzureStorageContainerStep.java

```diff
@@ -0,0 +1,142 @@
+  private static final Logger logger = LoggerFactory.getLogger(CreateAzureStorageContainerStep.class);
+  private final AzureConfiguration azureConfig;
+  private final CrlService crlService;
+  private final ControlledAzureStorageContainerResource resource;
+
+  public CreateAzureStorageContainerStep(
+      AzureConfiguration azureConfig,
+      CrlService crlService,
+      ControlledAzureStorageContainerResource resource) {
+    this.azureConfig = azureConfig;
+    this.crlService = crlService;
+    this.resource = resource;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    final AzureCloudContext azureCloudContext =
+        context
+            .getWorkingMap()
+            .get(ControlledResourceKeys.AZURE_CLOUD_CONTEXT, AzureCloudContext.class);
+    final StorageManager storageManager = crlService.getStorageManager(azureCloudContext, azureConfig);
+
+    try {
+      storageManager
+          .blobContainers()
+          .defineContainer(resource.getStorageContainerName())
+          .withExistingStorageAccount(azureCloudContext.getAzureResourceGroupId(), resource.getStorageAccountName())
+          .withPublicAccess(PublicAccess.NONE)
+          .withMetadata("workspaceId", resource.getWorkspaceId().toString())
+          .withMetadata("resourceId", resource.getResourceId().toString()).
+              create(
+                Defaults.buildContext(
+                  CreateStorageContainerRequestData.builder()
+                      .setName(resource.getStorageContainerName())
+                      .setStorageAccountName(resource.getStorageAccountName())
+                      .setResourceGroupName(azureCloudContext.getAzureResourceGroupId())
+                      .build()));
+
+    } catch (ManagementException e) {
+      logger.error(
+          "Failed to create the Azure storage container '{}' with storage account with the name '{}'. Error Code: {}",
+          resource.getStorageContainerName(),
+          resource.getStorageAccountName(),
+          e.getValue().getCode(),
+          e);
+
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+
+    return StepResult.getStepResultSuccess();
+  }
+
+  /**
+   * Deletes the storage container if the container is available. If the storage container is available
+   * and deletes fails, the failure is considered fatal and must looked into it.
+   *
+   * @param context
+   * @return Step result.
+   * @throws InterruptedException
+   */
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    final AzureCloudContext azureCloudContext =
+            context
+                    .getWorkingMap()
+                    .get(ControlledResourceKeys.AZURE_CLOUD_CONTEXT, AzureCloudContext.class);
+    final StorageManager storageManager = crlService.getStorageManager(azureCloudContext, azureConfig);
+
+    try {
+      storageManager
```

Comment:
🤔 We may need to add an additional authz check here on the storage account. I'm thinking through the scenario where I nefariously pass in a storage account name that is in the MRG but not "owned" by the workspace -- will this still succeed?

Reply:
This will get more complex when the storage account is in the landing zone.

Reply:
I guess I can check that the storage account is one of the controlled resources of the workspace. But I do wonder how all of this will change when the storage account moves to the landing zone.

Reply:
I think part of the problem is that `storageAccountName` is part of the request and cannot be trusted. When there is a landing zone it will dictate which storage account to use which means it should not be supplied in the request. I guess the question for now is can the workspace infer which storage account to use before landing zones come online? If they can, then do and remove it from the request. If not, well, I would encourage you to figure out an interim solution where they can :)

Reply:
I assume there is a way that I could query the workspace resources to see if a storage account already exists for the workspace. If one doesn't exist, we could error and say that a storage account must be created first. If one does exist, we would just use the first one we find?

I agree that not passing the storage account name would help protect the API from the storage account moving into the WLZ.

Reply:
How about we implement this current case by specifying the storage account resource id in the workspace to use for the create? We can look that up in the WSM metadata and do a Sam authz check on it to ensure the requesting user has access.

When landing zones are present, the landing zone manager will have to be able to provide the list of available landing zone storage accounts. When that is online, we can make a variant of the container create request that accepts the object id (or name) of the landing zone storage account to use. WSM can validate with LZ manager in the context of the LZ associated with the workspace.

I don't think we want to un-implement the case of a storage account allocated in the workspace.

Reply:
@ddietterich to be clear, are you suggesting that instead of the storage container create API taking the storage account name, it takes a storage account resource ID? Since the storage account name has to be unique, I think we could figure out the storage account resource ID from the name, though it would be a bit extra work.

Or are you suggesting when we create a storage account in the workspace, we store the storage account resource ID in the WSM metadata to use for future storage container create requests (and those do not specify a storage account)?

Reply:
I think that right now we only want to allow creation of containers from storage accounts in the workspace. When we have landing zones, we will also allow creation of containers from storage accounts in the landing zone.

Today when we make a storage account, we create a controlled resource in the workspace. It is scoped to the workspace and has a unique resource id. I'm suggesting you use that resource id of the storage account instead of the storage account name. The WSM metadata is setup to do lookups by resource id. You would have to craft a custom query on the database to look up by storage account name.

In the future when we have landing zones, we can amend the API to allow passing in some id for a landing zone storage account. AFAIK, the landing zone APIs have not been designed. At least I have not seen a doc.

Happy to have a short chat if this is not clear.

Reply:
Got it, I will change to the storage account resource ID.